### PR TITLE
Add monthly budget HUD to dashboard

### LINKED ISSUE #56: Upgrade the dashboard into a heads-up monthly budget HUD / Start of Phase 2
## Why it matters
This is the strongest UI and UX addition for final-project impact. It gives the app an immediately understandable budgeting-app feel the moment the dashboard loads.
## Scope
Add a dashboard HUD section that prominently shows left to spend this month, a budget progress bar, days left in the month, optional daily allowance based on remaining budget divided by days left, over or under budget status, and a compact highlight such as top category pressure or strongest overspend area. The module should prioritize glanceable status and stronger visual urgency when the user is over budget or close to the limit.
## Acceptance Criteria  
- A signed-in user can understand their monthly budget position from the top of the dashboard without scrolling.
- The HUD shows remaining budget and progress visually, not just as text.
- If a budget exists, the HUD can show days left and a daily-allowance style figure.
- The HUD gracefully handles no-budget, under-budget, and over-budget states.
- Over-budget or high-risk states read clearly at a glance without requiring users to parse dense copy.
- The module remains readable on mobile and in both themes.
## Story Points 
3
## Dependencies
Depends on #43

### PR DESCRIPTION
**Pull request by adillpickles**
## Description
Adds a heads-up monthly budget HUD to the dashboard so users can understand their monthly budget position immediately from the top of the page. This refactors the existing dashboard hero into a more glanceable budget module with a visual progress bar, days left in the month, daily allowance, net-this-month, and a compact category pressure / overspend highlight, while keeping the existing trend chart as secondary context.

## Related User Story / Issue
Fixes #56

## Type of Change
- [x] New feature
- [ ] Bug fix
- [ ] Backend / API change
- [x] UI change
- [ ] Database change
- [ ] Documentation
- [ ] Refactor

## How Has This Been Tested?
- [ ] GitHub Actions / CI tests passed
- [x] Manual testing performed
- [ ] Not tested locally

### Test Notes
Verified locally with:
- `npm test -- --runInBand dashboard.test.js`
- `npm test -- --runInBand planner.test.js insights.test.js budget.test.js`
- `npm run build`

Added regression coverage for HUD derivation logic and state handling, including:
- no-budget state
- on-track / near-limit / over-budget states
- days-left and daily-allowance calculations
- category pressure highlight priority and fallback behavior

## UI Changes (if applicable)
- [ ] No UI changes
- [x] UI updated (add screenshots if needed)

## BEFORE
image
image

## AFTER
image

## Notes for Reviewers
This stays intentionally scoped to the dashboard/UI layer and does not change the budget summary API contract. The HUD derives its extra state client-side from existing summary, expenses, and income data already available on `main`.

This also does not attempt to fully complete the broader cross-screen consistency work from #43; it only implements the dashboard portion needed for #56.

## Checklist
- [x] Linked to a user story or issue
- [x] Code builds / tests pass in CI
- [x] Ready for review

### CODE REVIEW

File: nextjs/src/components/dashboard-view.js [hunk 1]
```diff
@@ -77,56 +77,202 @@ export function getBudgetHintText(summary) {
...
+      monthState,
+      daysRemaining: null,
+      dailyAllowance: null,
+      isOverBudget: false,
+      isNearLimit: false,
+      hasBudget: false,
+      metrics: [
+        { label: 'Spent', value: formatCurrency(spent), hint: 'Current month' },
+        { label: 'Income', value: formatCurrency(income), hint: 'Current month' },
+        { label: 'Net this month', value: formatCurrency(net), hint: 'Income minus spend' },
```
[the remark this answers is not in the document]
Comment (adillpickles) [reply nextjs/src/components/dashboard-view.js]: Good catch. I went ahead and changed the HUD loading state so those metrics show `--` instead of briefly rendering misleading $0 values before the summary data arrives. I also added a focused dashboard regression test for it.

File: nextjs/src/components/dashboard-view.js [hunk 2]
```diff
@@ -77,56 +77,202 @@ export function getBudgetHintText(summary) {
...
+      daysRemaining: null,
+      dailyAllowance: null,
+      isOverBudget: false,
+      isNearLimit: false,
+      hasBudget: false,
+      metrics: [
+        { label: 'Spent', value: '--', hint: 'Current month' },
+        { label: 'Income', value: '--', hint: 'Current month' },
+        { label: 'Net this month', value: '--', hint: 'Income minus spend' },
+      ],
```
[the remark this answers is not in the document]
Comment (adillpickles) [reply nextjs/src/components/dashboard-view.js:201]: Good polish catch. I updated the loading/no-budget HUD states so they now render a fourth metric as well, which keeps the grid consistent and avoids the empty-column layout.

File: nextjs/src/components/dashboard-view.js [hunk 3]
```diff
@@ -77,56 +77,202 @@ export function getBudgetHintText(summary) {
...
+  const monthLength = new Date(Date.UTC(
+    monthDate.getUTCFullYear(),
+    monthDate.getUTCMonth() + 1,
+    0
+  )).getUTCDate()
+  const currentDate = referenceDate instanceof Date ? referenceDate : new Date(referenceDate)
+  const isCurrentMonth = (
+    monthDate.getUTCFullYear() === currentDate.getUTCFullYear()
+    && monthDate.getUTCMonth() === currentDate.getUTCMonth()
+  )
```
[the remark this answers is not in the document]
Comment (adillpickles) [reply nextjs/src/components/dashboard-view.js]: standardized the HUD month logic onto the same local month basis used elsewhere in the dashboard, and added regression coverage for the month-boundary case so `daysRemaining` / daily allowance stay consistent.

File: nextjs/__tests__/frontend/dashboard.test.js
Comment (zakariac22) [inline nextjs/__tests__/frontend/dashboard.test.js:1]: When running `npx jest --coverage __tests__/dashboard.test.js` it gives statement coverage of 37.5%. Can you increase coverage for this to ~70%? 

Also, is it possible to add integration tests for this? It would be meaningful, as there are direct API calls on the dashboard-view.js component.

Comment (adillpickles) [reply nextjs/__tests__/frontend/dashboard.test.js:1]: That’s fair. I’ll do a follow-up pass on the dashboard tests and try to raise coverage with more targeted cases around the new HUD behavior. I’ll also look at whether there’s a clean integration style test we can add without turning this into a broader test rewrite.

Comment (adillpickles) [reply nextjs/__tests__/frontend/dashboard.test.js:1]: I did a follow-up testing pass on the dashboard PR. Coverage for `dashboard-view.js` is now at 93.05% statements, 72.24% branches, 97.01% functions, and 95.34% lines. I also added integration-style dashboard tests around the direct API behavior, including live fetch/render, partial failure, unauthorized redirect, and budget-sheet save flow.

File: nextjs/src/components/dashboard-view.js [hunk 4]
```diff
@@ -77,56 +77,210 @@ export function getBudgetHintText(summary) {
...
+  const monthLength = new Date(monthYear, monthIndex + 1, 0).getDate()
+  const isCurrentMonth = getCurrentMonthStart(currentDate) === month
+  const fallbackObservedDays = Number.isFinite(observedDayCount)
+    ? Math.max(0, Math.floor(observedDayCount))
+    : 0
+  const activeDay = isCurrentMonth
+    ? Math.min(Math.max(currentDate.getDate(), 1), monthLength)
+    : Math.min(Math.max(fallbackObservedDays || monthLength, 1), monthLength)
+  const daysRemaining = isCurrentMonth || fallbackObservedDays
+    ? Math.max(monthLength - activeDay + 1, 0)
```
[the remark this answers is not in the document]
Comment (adillpickles) [reply nextjs/src/components/dashboard-view.js]: I tightened the date handling so invalid `referenceDate` values now fall back safely instead of leaking `NaN` into the HUD state.

File: nextjs/__tests__/dashboard.test.js [hunk 5]
```diff
@@ -177,3 +296,447 @@ describe('buildDerivedCategoryCards', () => {
...
+      1,
+      '/api/budget/summary?month=2026-04-01',
+      expect.objectContaining({ accessToken: 'test-token', signal: expect.any(AbortSignal) })
+    )
+    expect(text).toContain('Live')
+    expect(text).toContain('Near limit')
+    expect(text).toContain('$140 left')
+    expect(text).toContain('Top category pressure')
+    expect(text).toContain('Grocer')
+    expect(text).toContain('Paycheck')
```
[the remark this answers is not in the document]
Comment (adillpickles) [reply nextjs/__tests__/dashboard.test.js]: I reworked the dashboard test setup so the suite now uses one frozen deterministic runtime date instead of scattered hardcoded month assumptions.

File: nextjs/__tests__/dashboard.test.js [hunk 6]
```diff
@@ -177,3 +296,447 @@ describe('buildDerivedCategoryCards', () => {
...
+    apiPost.mockResolvedValueOnce({})
+
+    const renderer = await renderDashboard()
+
+    await act(async () => {
+      getButtonByText(renderer, 'Edit budget').props.onClick()
+    })
+    expect(getRenderedText(renderer)).toContain('Monthly spending limit for')
+    expect(getRenderedText(renderer)).toContain('2026-04-01')
+
```
[the remark this answers is not in the document]
Comment (adillpickles) [reply nextjs/__tests__/dashboard.test.js]: Yep, fixed this too. The budget-sheet expectation now comes from the same deterministic test month as the rest of the dashboard suite, so it shouldn’t drift over time anymore.

File: nextjs/__tests__/dashboard.test.js [hunk 7]
```diff
@@ -24,20 +79,107 @@ jest.mock('@/lib/financeVisuals', () => ({
...
+  return {
+    month: TEST_CURRENT_MONTH,
+    ...overrides,
+  }
+}
+
+const TEST_NOW = new Date('2026-03-21T12:00:00Z')
+const TEST_CURRENT_MONTH = getMonthStartString(TEST_NOW)
+const TEST_NEXT_MONTH_DATE = new Date('2026-04-21T12:00:00Z')
+const TEST_NEAR_BOUNDARY_CURRENT_MONTH_DATE = new Date('2026-03-01T00:30:00-05:00')
```
[the remark this answers is not in the document]
Comment (adillpickles) [reply nextjs/__tests__/dashboard.test.js]: i tightened the dashboard date-based test setup so it stays deterministic across machines and timezones instead of depending on offset-based ISO fixtures.

File: nextjs/src/components/dashboard-view.js [hunk 8]
```diff
@@ -77,56 +77,215 @@ export function getBudgetHintText(summary) {
...
+}
+
+function getSafeReferenceDate(referenceDate) {
+  const parsedReferenceDate = referenceDate instanceof Date ? referenceDate : new Date(referenceDate)
+  return Number.isNaN(parsedReferenceDate.getTime()) ? new Date() : parsedReferenceDate
+}
+
+export function getMonthProgressState(month, { observedDayCount = 0, referenceDate = new Date() } = {}) {
+  const monthMatch = typeof month === 'string'
+    ? month.match(/^(\d{4})-(\d{2})-\d{2}$/)
```
[the remark this answers is not in the document]
Comment (adillpickles) [reply nextjs/src/components/dashboard-view.js]: I made the month parsing stricter so invalid values now return the same safe empty state instead of partially parsing into misleading output.

File: nextjs/src/components/dashboard-view.js [hunk 9]
```diff
@@ -77,56 +77,215 @@ export function getBudgetHintText(summary) {
...
+      daysRemaining: 0,
+      isCurrentMonth: false,
+    }
+  }
+
+  const currentDate = getSafeReferenceDate(referenceDate)
+  const monthYear = Number(monthMatch[1])
+  const monthIndex = Number(monthMatch[2]) - 1
+  const monthLength = new Date(monthYear, monthIndex + 1, 0).getDate()
+  const isCurrentMonth = getCurrentMonthStart(currentDate) === month
```
[the remark this answers is not in the document]
Comment (adillpickles) [reply nextjs/src/components/dashboard-view.js]: The month handling now normalizes valid inputs before comparing, so `isCurrentMonth` no longer depends on the caller already passing a month-start string.

File: nextjs/__tests__/frontend/dashboard.test.js
Comment (zakariac22) [reply nextjs/__tests__/frontend/dashboard.test.js:1]: For integration testing, I would advise using the React Testing Library to render the component. You can mock the API response and assert UI updates were done properly.

This will let you simulate real user interactions and check the dashboard behaves correctly. You can keep using Jest for testing, React Testing Library is only used for renders.

Comment (adillpickles) [reply nextjs/__tests__/frontend/dashboard.test.js:1]: fixed in latest commit

File: nextjs/src/components/dashboard-view.js [hunk 10]
```diff
@@ -531,6 +777,52 @@ export default function DashboardView() {
...
+            aria-label="Monthly budget progress"
+            className="budget-progress budget-hero__progress"
+            role="progressbar"
+            aria-valuemin={0}
+            aria-valuemax={100}
+            aria-valuenow={Math.round(hudState.progressPercentage)}
+            aria-valuetext={hudState.progressLabel}
+          >
+            <span
+              className={`budget-progress__fill budget-progress__fill--${hudState.tone}`}
```
[the remark this answers is not in the document]
Comment (adillpickles) [reply nextjs/src/components/dashboard-view.js:804]: I added an explicit `budget-progress__fill--positive` rule so the positive state no longer depends on the base fill styling implicitly.

File: nextjs/src/components/dashboard-view.js [hunk 11]
```diff
@@ -470,14 +714,17 @@ export default function DashboardView() {
...
     ? demoBudgetTrend
     : buildMonthlySpendTrend(liveState.expenses, currentMonth)
-  const projectedTrendPoints = buildProjectedTrend(chartMonth, heroState.budget, trendPoints.length)
-  const chartCeiling = getTrendCeiling(trendPoints, projectedTrendPoints, heroState.budget)
+  const hudState = getBudgetHudModel(summary, {
+    month: chartMonth,
+    observedDayCount: trendPoints.length,
+  })
+  const pressureHighlight = getBudgetPressureHighlight(summary, currentMonthExpenses)
+  const projectedTrendPoints = buildProjectedTrend(chartMonth, hudState.budget, trendPoints.length)
```
[the remark this answers is not in the document]
Comment (adillpickles) [reply nextjs/src/components/dashboard-view.js:732]: I refactored the fallback path so the derived category cards are computed once per render and then reused for both the category cards and the pressure highlight.

File: nextjs/__tests__/frontend/dashboard.test.js
Comment (zakariac22) [inline nextjs/__tests__/frontend/dashboard.test.js:1]: Move this file to the __tests__/frontend folder. Otherwise, this seems ready for approval

Comment (adillpickles) [reply nextjs/__tests__/frontend/dashboard.test.js:1]: Done in the latest commit, I moved the dashboard test into `__tests__/frontend` and kept the RTL based integration coverage there.

File: nextjs/src/components/dashboard-view.js [hunk 12]
```diff
@@ -454,6 +699,8 @@ export default function DashboardView() {
     : buildActivityFeed(liveState.expenses, liveState.income)
   const recentActivity = activity.slice(0, PREVIEW_LIMIT)
   const recentIncome = activity.filter((entry) => entry.kind === 'income').slice(0, INCOME_LIMIT)
+  const chartMonth = isSampleMode ? DEMO_MONTH : summary?.month || currentMonth
+  const derivedCategoryCards = isSampleMode ? [] : buildDerivedCategoryCards(currentMonthExpenses)
```
[the remark this answers is not in the document]
Comment (adillpickles) [reply nextjs/src/components/dashboard-view.js]: tightened that path so the derived fallback cards are only computed when they’re actually needed, instead of being built on live renders that already have budgeted category statuses.

**Review by zakariac22 — APPROVED** [review]
